test/client/fixtures/sandbox/event/event-simulator-test.js

```diff
@@ -475,6 +475,81 @@ if (!browserUtils.isFirefox) {
+            which:   e.which
+        });
+    };
+
+    var fillExpectedLog = function (type, button, buttons, which) {
+        expectedLog[type] = expectedLog[type] || [];
+
+        // NOTE: Safari doesn't support buttons property
+        expectedLog[type].push({
+            button:  button,
+            buttons: getButtonsProperty(buttons),
+            which:   which
+        });
+    };
+
+    fillExpectedLog('click', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('dblclick', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('contextmenu', eventUtils.BUTTON.right, eventUtils.BUTTONS_PARAMETER.rightButton, eventUtils.WHICH_PARAMETER.rightButton);
+    fillExpectedLog('mousedown', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mousedown', eventUtils.BUTTON.right, eventUtils.BUTTONS_PARAMETER.rightButton, eventUtils.WHICH_PARAMETER.rightButton);
+    fillExpectedLog('mouseup', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mouseup', eventUtils.BUTTON.right, eventUtils.BUTTONS_PARAMETER.rightButton, eventUtils.WHICH_PARAMETER.rightButton);
+    fillExpectedLog('mouseover', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mouseout', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mouseenter', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mouseleave', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog('mousemove', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, eventUtils.WHICH_PARAMETER.leftButton);
+    fillExpectedLog(
+        'mousemove',
+        eventUtils.BUTTON.left,
+        eventUtils.BUTTONS_PARAMETER.noButton,
+        browserUtils.isWebKit ? eventUtils.WHICH_PARAMETER.noButton : eventUtils.WHICH_PARAMETER.leftButton);
+
+    domElement.addEventListener('click', eventHandler);
+    domElement.addEventListener('dblclick', eventHandler);
+    domElement.addEventListener('contextmenu', eventHandler);
+    domElement.addEventListener('mousedown', eventHandler);
+    domElement.addEventListener('mouseup', eventHandler);
+    domElement.addEventListener('mouseover', eventHandler);
+    domElement.addEventListener('mouseout', eventHandler);
+    domElement.addEventListener('mouseenter', eventHandler);
+    domElement.addEventListener('mouseleave', eventHandler);
+    domElement.addEventListener('mousemove', eventHandler);
+
+    eventSimulator.click(domElement);
```

Comment:
Let reorganize this test in data driven style, e.g.:
```js
const testData = [
    {
        eventSimulator: () => eventSimulator.dblclick(domElement),
        expectedLog: getExpectedLog('click', eventUtils.BUTTON.left, eventUtils.BUTTONS_PARAMETER.leftButton, 
        eventUtils.WHICH_PARAMETER.leftButton);
    },
...
testData.forEach(function (item) {
     item.eventSimulator();
     deepEqual(actualLog, item.expectedLog);
     clearLogs();
});
}
```
IMHO It's easier to understand and debug. What do you think?

Reply:
Agree. I'organize the test in this manner